File: ΕΧ2/Arrange_java/Arrange.java
package ΕΧ2.Arrange_java;
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Scanner;

public class Arrange {
    static class NodeResult {
        TreeNode node;
        int index;

        public NodeResult(TreeNode node, int index) {
            this.node = node;
            this.index = index;
        }
    }

    static TreeNode swapN(TreeNode n) {
        TreeNode temp = n.left;
        n.left = n.right;
        n.right = temp;
        return n;
    }

    static boolean isLeaf(TreeNode n) {
        return n.left == null && n.right == null;
    }

    static int minleaf(TreeNode n) {
        if (isLeaf(n)) return n.info;
        if (n.left == null) return Math.min(n.info, minleaf(n.right));
        if (n.right == null) return Math.min(n.info, minleaf(n.left));
        return Math.min(minleaf(n.left), minleaf(n.right));
    }

    static void inorder(TreeNode n) {
        if (n == null) return;
        inorder(n.left);
        System.out.print(n.info + " ");
        inorder(n.right);
    }

    static NodeResult prein(int[] inf, int index) {
        if (inf[index] == 0) return new NodeResult(null, index + 1);
        TreeNode p = new TreeNode(inf[index]);
        NodeResult leftResult = prein(inf, index + 1);
        p.left = leftResult.node;
        NodeResult rightResult = prein(inf, leftResult.index);
        p.right = rightResult.node;
        return new NodeResult(p, rightResult.index);
    }

    static TreeNode fix(TreeNode n) {
        if (n == null) return null;
        if (isLeaf(n)) return n;
        if (n.left == null && n.right != null) {
            if (n.info > minleaf(n.right)) {
                n = swapN(n);
                n.left = fix(n.left);
                n.right = null;
            } else {
                n.right = fix(n.right);
                n.left = null;
            }
            return n;
        }
        if (n.right == null && n.left != null) {
            if (n.info < minleaf(n.left)) {
                n = swapN(n);
                n.right = fix(n.right);
                n.left = null;
            } else {
                n.left = fix(n.left);
                n.right = null;
            }
            return n;
        }
        if (minleaf(n.right) < minleaf(n.left)) n = swapN(n);
        n.left = fix(n.left);
        n.right = fix(n.right);
        return n;
    }

    public static void main(String[] args) {
        if (args.length == 0) {
            System.out.println("Usage: java Main <input_file>");
            return;
        }

        try {
            Scanner scanner = new Scanner(new File(args[0]));
            int n = scanner.nextInt();
            int[] arr = new int[2 * n + 1];
            int i = 0;

            while (i < 2 * n + 1) {
                arr[i++] = scanner.nextInt();
            }

            NodeResult rootResult = prein(arr, 0);
            TreeNode root = rootResult.node;
            root = fix(root);
            inorder(root);
            System.out.println();
            scanner.close();
        } catch (FileNotFoundException e) {
            System.out.println("File not found: " + args[0]);
        }
    }
}

File: ΕΧ1/arrange.cpp
#include <fstream>
using namespace std;
struct node { /* recursive node definition*/
    int info;
    node *left;
    node *right;
};

node* swapN(node* n){ /*swaps the children of a node n*/
    node* temp = n->left;
    n->left = n->right;
    n->right = temp;
    return n;
}


bool isLeaf(node* n){ /*checks if the node is a leaf*/
    return(n->left == nullptr && n->right == nullptr) ;
}
int minleaf(node * n){ /* checks the children of a node to see which one has the smallest numbered leaf*/
    if(isLeaf(n)) return n->info;
    if( n->left ==nullptr) return min(n-> info, minleaf(n->right));
    if(n->right == nullptr) return min(n->info, minleaf(n->left));
    return min(minleaf(n->left), minleaf(n->right));

}

void inorder(node* n){ /*classic inorder function*/
    if(n == nullptr)return;
    inorder(n->left);
    printf("%d ", n->info);
    inorder(n->right);
}
node* prein(node *n , int *(&inf)){ /* recursive insert function that takes an array with the elements in preorder as input and then creates a tree*/
    node* p = new node;
    if(*inf == 0){ return nullptr;} 
    p->info = *inf;

    p->left = prein(p, (++inf));
    p->right = prein(p, (++inf));
    return p;

}
node* fix(node* n){ /* this is the function that balances the tree so that when you call inorder it is in the correct lexicographic order*/
    if(n == nullptr) return nullptr;
    if(n->left == nullptr && n->right == nullptr) return n;
    if(n->left == nullptr && n->right != nullptr){
        if(n-> info > minleaf(n->right)) {
            n = swapN(n);
            n->left = fix(n->left);
            n->right = nullptr;
        }
        else {
            n->right = fix(n->right);
            n->left = nullptr;
        }	
        return n;
    }
    if(n->right == nullptr && n->left != nullptr) {
        if(n-> info < minleaf(n->left)) {
            n = swapN(n);
            n->right = fix(n->right);
            n->left = nullptr;
        }
        else{
            n->left = fix(n->left);
            n->right = nullptr;
        }
        return n;
    }
    else{
        if(minleaf(n->right)< minleaf(n->left)) n = swapN(n);
        n->left = fix(n->left);
        n->right = fix(n->right);
        return n;
    }

    return n;
}


int main (int argc, char ** argv){
    int n = 0;

    ifstream infile;
    infile.open(argv[1]);
    infile>>n;

    int* arr = new int [2*n+1];
    int i = 0, info =0;

    while(i < 2*n+1){

        infile>>info;

        arr[i++] = info;        
    }
    node* root = new node;
    root->left = nullptr;
    root->right = nullptr;
    root->info = 1;
    int * aux =arr;

    root = prein(root, aux);

    root = fix(root);

    inorder(root);
    printf("\n");
    delete [] arr; 
}